core/server/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -0,0 +1,338 @@
+
+/**
+ * This class handles {@link alluxio.proto.dataserver.Protocol.WriteRequest}s.
+ *
+ * Protocol: Check {@link alluxio.client.block.stream.NettyPacketWriter} for more information.
+ * 1. The netty channel handler streams packets from the channel and buffers them. The netty
+ *    reader is paused if the buffer is full by turning off the auto read, and is resumed when
+ *    the buffer is not full.
+ * 2. The {@link PacketWriter} polls packets from the buffer and writes to the block worker. The
+ *    writer becomes inactive if there is nothing on the buffer to free up the executor. It is
+ *    resumed when the buffer becomes non-empty.
+ * 3. When an error occurs, the channel is closed. All the buffered packets are released when the
+ *    channel is deregistered.
+ */
+@NotThreadSafe
+public abstract class DataServerWriteHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.WORKER_NETWORK_NETTY_WRITER_BUFFER_SIZE_PACKETS);
+
+  /** The executor service to run the {@link PacketWriter}s. */
+  private final ExecutorService mPacketWriterExecutor;
+
+  private ReentrantLock mLock = new ReentrantLock();
+  /** The buffer for packets read from the channel. */
+  @GuardedBy("mLock")
+  private Queue<ByteBuf> mPackets = new LinkedList<>();
+  /** Set to true if the packet writer is active. */
+  @GuardedBy("mLock")
+  private boolean mPacketWriterActive = false;
+
+  protected volatile WriteRequestInternal mRequest = null;
+
+  protected abstract class WriteRequestInternal implements Closeable {
+    // This ID can either be block ID or temp UFS file ID.
+    public long mId = -1;
+    public long mSessionId = -1;
+  }
+
+  /**
+   * The next pos to queue to the buffer. Updated by the packet reader. Mostly used to validate
+   * the validity of the request.
+   */
+  private volatile long mPosToQueue = 0;
+  /**
+   * The next pos to write to the block worker. Updated by the packet writer.
+   */
+  protected volatile long mPosToWrite = 0;
+
+  /**
+   * Creates an instance of {@link DataServerWriteHandler}.
+   *
+   * @param executorService the executor service to run {@link PacketWriter}s
+   */
+  public DataServerWriteHandler(ExecutorService executorService) {
+    mPacketWriterExecutor = executorService;
+  }
+
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object object) throws Exception {
+    if (!acceptMessage(object)) {
+      ctx.fireChannelRead(object);
+      return;
+    }
+
+    RPCProtoMessage msg = (RPCProtoMessage) object;
+    initializeRequest(msg);
+
+    // Validate msg and return error if invalid. Init variables if necessary.
+    String error = validateRequest(msg);
+    if (!error.isEmpty()) {
+      replyError(ctx.channel(), Protocol.Status.Code.INVALID_ARGUMENT, error, null);
+      return;
+    }
+
+    mLock.lock();
```

Comment:
NIT: this may require many change, feel free to make it a TODO: this PR has this pattern repeated many times like 
```java
mLock.lock();
try {
} finally {
  mLock.unlock();
}
```
Can we use the following pattern to simply the code and improve readability?
```java
try (LockResource lr = new LockResource(mLock)) {
}
```